src/Umbraco.Core/Services/NotificationService.cs

```diff
@@ -55,38 +55,54 @@ public NotificationService(IDatabaseUnitOfWorkProvider provider, IUserService us
+            // do not load *all* users in memory at once
+            // do not load notifications *per user* (N+1 select)
+            // cannot load users & notifications in 1 query (combination btw User2AppDto and User2NodeNotifyDto)
+            // => get batches of users, get all their notifications in 1 query
+            // re. users:
+            //  users being (dis)approved = not an issue, filtered in memory not in SQL
+            //  users being modified or created = not an issue, ordering by ID, as long as we don't *insert* low IDs
+            //  users being deleted = not an issue for GetNextUsers
+            var id = 0;
+            var nodeIds = content.Path.Split(',').Select(int.Parse).ToArray();
+            const int pagesz = 400; // load batches of 400 users
+            do
             {
-                var userNotifications = GetUserNotifications(u, content.Path);
-                var notificationForAction = userNotifications.FirstOrDefault(x => x.Action == action);
-                if (notificationForAction == null) continue;
-
-                if (allVersions == null) // lazy load
-                    allVersions = _contentService.GetVersions(entity.Id).ToList();
+                // users are returned ordered by id, notifications are returned ordered by user id
+                var users = ((UserService) _userService).GetNextUsers(id, pagesz).Where(x => x.IsApproved).ToList();
+                var notifications = GetUsersNotifications(users.Select(x => x.Id), action, nodeIds, Constants.ObjectTypes.DocumentGuid).ToList();
+                if (notifications.Count == 0) break;
 
-                try
+                var i = 0;
+                foreach (var user in users)
                 {
-                    SendNotification(operatingUser, u, content, allVersions, 
-                        actionName, http, createSubject, createBody);
+                    // continue if there's no notification for this user
+                    if (notifications[i].UserId != user.Id) continue; // next user
 
-                    _logger.Debug<NotificationService>(string.Format("Notification type: {0} sent to {1} ({2})",
-                        action, u.Name, u.Email));
-                }
-                catch (Exception ex)
-                {
-                    _logger.Error<NotificationService>("An error occurred sending notification", ex);
+                    // lazy load all versions
+                    if (allVersions == null) allVersions = _contentService.GetVersions(entity.Id).ToList();
+
+                    // queue notification
+                    var req = CreateNotificationRequest(operatingUser, user, content, allVersions, actionName, http, createSubject, createBody);
+                    Enqueue(req);
+
+                    // skip other notifications for this user
+                    while (i < notifications.Count && notifications[i++].UserId == user.Id) ;
```

Comment:
@zpqrtbnk  Shouldn't this be `++i` instead of `i++` ? Otherwise the first iteration will only be using i as it is 'now' to compare the UserId's and then it will increment it. The 2nd iteration it will have now incremented it and will compare the UserId's but then it will increment it again  when the boolean no longer matches. This means that an `i` iteration has bee skipped.